### PR TITLE
[16.0][ADD] account_invoice_edifact

### PR DESCRIPTION
**Pull request by thienvh332**
This new module will support generate customer invoice with EDIFACT/D96A format.

Depends on:

- https://github.com/OCA/edi/pull/853

### CODE REVIEW

File: account_invoice_edifact/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+from datetime import datetime
+
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    def edifact_export_invoice(self):
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]: It's just generating data, not really exporting

```suggestion
    def edifact_invoice_generate_data(self):
```

File: account_invoice_edifact/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+            interchange.add_segment(edifact_model.create_segment(*segment))
+        return interchange.serialize()
+
+    def _edifact_export_invoice_interchange(self):
+        self.ensure_one()
+        id_number = self.env["res.partner.id_number"]
+        sender = id_number.search(
+            [("partner_id", "=", self.invoice_user_id.partner_id.id)]
+        )
+        recipient = id_number.search([("partner_id", "=", self.partner_id.id)])
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
        sender = id_number.search(
            [("partner_id", "=", self.invoice_user_id.partner_id.id)],
            limit=1
        )
        recipient = id_number.search([("partner_id", "=", self.partner_id.id)], limit=1)
```

File: account_invoice_edifact/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+        header = self._edifact_export_invoice_header()
+        product, vals = self._edifact_export_invoice_product()
+        summary = self._edifact_export_invoice_summary(vals)
+        lines += header + product + summary
+        for segment in lines:
+            interchange.add_segment(edifact_model.create_segment(*segment))
+        return interchange.serialize()
+
+    def _edifact_export_invoice_interchange(self):
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
    def _edifact_invoice_get_interchange(self):
```

File: account_invoice_edifact/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+            raise UserError(_("Partner is not allowed to use the feature."))
+        sender_edifact = [sender.name, "14"]
+        recipient_edifact = [recipient.name, "14"]
+        syntax_identifier = ["UNOC", "3"]
+
+        return self.env["base.edifact"].create_interchange(
+            sender_edifact, recipient_edifact, self.id, syntax_identifier
+        )
+
+    def _edifact_export_invoice_header(self):
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
    def _edifact_invoice_get_header(self):
```

File: account_invoice_edifact/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        source_orders = self.line_ids.sale_line_ids.order_id
+        today = datetime.now().date().strftime("%Y%m%d")
+        id_number = self.env["res.partner.id_number"]
+        buyer = self.partner_id
+
+        seller = self.invoice_user_id.partner_id
+        seller_id_number = id_number.search([("partner_id", "=", seller.id)])
+
+        shipper = self.partner_shipping_id
+        shipper_id_number = id_number.search([("partner_id", "=", shipper.id)])
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
        seller_id_number = id_number.search([("partner_id", "=", seller.id)], limit=1)

        shipper = self.partner_shipping_id
        shipper_id_number = id_number.search([("partner_id", "=", shipper.id)], limit=1)
```

File: account_invoice_edifact/models/account_move.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+        payment_term = min(self.invoice_payment_term_id.line_ids, default=None)
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]: I'm no sure to understand this usage of `min` here

File: account_invoice_edifact/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            # 35: Delivery date/time, actual
+            (
+                "DTM",
+                [
+                    "35",
+                    max(
+                        (
+                            picking.date_done.date().strftime("%Y%m%d")
+                            for order in source_orders
+                            for picking in order.picking_ids
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py:159]: shouldn't you filter on the delivery only?

File: account_invoice_edifact/models/account_move.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            ("FTX", "PMD", "", "", ""),
+            # Terms of payments
+            ("FTX", "AAB", "", "", "30 jours net"),
+            # Delivery note number
+            ("RFF", ["DQ", self.id]),
+            # Reference date/time
+            # TODO: fixed value for now, to be clarified
+            ("DTM", ["171", "99991231", "102"]),
+            # Invoice information
+            (
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]: maybe these parts can be split to their own methods? Eg: `_edifact_invoice_get_[buyer|seller|you-name-it]`.

File: account_invoice_edifact/models/account_move.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            ("PAT", "20"),
+            # Penalty percentage
+            ("PCD", "15", "0"),  # TODO: check value this again later
+            # Allowance percentage
+            ("PCD", "1", "0", "13"),  # TODO: check value this again later
+            # Allowance or charge amount
+            ("MOA", "8", "0"),  # TODO: check value this again later
+        ]
+
+    def _edifact_export_invoice_product(self):
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
    def _edifact_invoice_get_product(self):
```

File: account_invoice_edifact/models/account_move.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        self.ensure_one()
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]: As these methods are a private api meant to be used inside the main one to me is useless to keep ensuring that you get one record only... not a blocker tho.

File: account_invoice_edifact/models/account_move.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+        number = 0
+        segments = []
+        vals = {}
+        tax = {}
+        for line in self.line_ids:
+            if line.display_type == "product":
```
Comment (simahawk) [inline account_invoice_edifact/models/account_move.py]:
```suggestion
            if line.display_type != "product":
                continue
```
and you save a looooot of indentation in the rest of the method

File: account_invoice_edifact/tests/test_account_move.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
+# Copyright 2023 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests.common import TransactionCase
+
+
+class TestEdifactInvoice(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
Comment (simahawk) [inline account_invoice_edifact/tests/test_account_move.py:10]: pls disable mail tracking globally

**Review by simahawk — CHANGES_REQUESTED** [review]

File: account_invoice_edifact/models/account_move.py (hunk 7 above)
Comment (thienvh332) [reply account_invoice_edifact/models/account_move.py:159]: On the document they sent there was no DTM+35, so I relied on the EDIFACT convention. In EDIFACT DTM+35 is Delivery date/time, actual. So I decided to get picking.date_done. If you have any suggestions I will fix it.

File: account_invoice_edifact/tests/test_account_move.py (hunk 12 above)
Comment (thienvh332) [reply account_invoice_edifact/tests/test_account_move.py:10]: Thanks @simahawk , I updated

**Review by simahawk — APPROVED** [review]

**Pull request merged**